### PR TITLE
5 jackyao harpoongun

### PR DESCRIPTION
**Pull request by j-ackyao**
- Added `CollisionUtils`
- Added `IntroSkill` to `WeaponBase`, which should be invoked upon weapon switching
- `HarpoonGun` base structure, 
- Still to-be implemented features:
    - `UseSkill`
        - Blocked by `PlayerController`, who needs an interface to allow player manipulation from other classes
    - `IntroSkill`
        - Blocked by `PlayerController`, which has yet to implement weapon switching

### CODE REVIEW

File: Assets/Scripts/Weapon/HarpoonGun/HarpoonSpear.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    public Collider2D Collider2D;
+    public Rigidbody2D Rigidbody2D;
+
+    private bool dropped;
+    private HarpoonGun gun;
+
+    // Tracking enemy
+    private Enemy taggedEnemy;
+    public Enemy TaggedEnemy {
+        get { return taggedEnemy; }
```
Comment (anjx123) [inline Assets/Scripts/Weapon/HarpoonGun/HarpoonSpear.cs:23]: Could collapse this into one property getter with a autogenerated backing field

public Enemy TaggedEnemy {get; private set;}

File: Assets/Scripts/Weapon/HarpoonGun/HarpoonGun.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+{
+
+    [Header("HarpoonGun Stats")]
+    [SerializeField] public int harpoons = 3;
+    [SerializeField] public float harpoonVelocity = 20;
+    [SerializeField] public float harpoonSpearGravityScale = 1;
+    [SerializeField] public GameObject harpoonSpearPrefab;
+
+    // Assuming harpoons isn't too great, otherwise will use HashSet
+    LinkedList<HarpoonSpear> spears = new LinkedList<HarpoonSpear>();
```
Comment (anjx123) [inline Assets/Scripts/Weapon/HarpoonGun/HarpoonGun.cs:15]: Linked list skull emoji

Comment (anjx123) [reply Assets/Scripts/Weapon/HarpoonGun/HarpoonGun.cs:15]: Use a Queue apparent type, and linked list actual type :)

File: Assets/Scripts/Weapon/HarpoonGun/HarpoonGun.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    public override void Attack()
+    {
+        if (spears.Count >= harpoons) {
+            // Do nothing
+            return;
+        }
+
+        GameObject spearObject = Instantiate(harpoonSpearPrefab, transform.position, transform.rotation);
```
Comment (anjx123) [inline Assets/Scripts/Weapon/HarpoonGun/HarpoonGun.cs:24]: I would consider using object pooling for projectiles, but maybe in the future

File: Assets/Scripts/Weapon/HarpoonGun/HarpoonGun.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    public override void UseSkill()
+    {
+        if (spears.Count == 0) {
+            return;
+        }
+
+        // Find the next most recent spear that has tagged enemy
+        Enemy tagged = null;
```
Comment (anjx123) [inline Assets/Scripts/Weapon/HarpoonGun/HarpoonGun.cs:37]: spears.Where(spear => spear.TaggedEnemy != null)

File: Assets/Scripts/Weapon/HarpoonGun/HarpoonSpear.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    public void OnDisable() {
+        Enemy.Death -= HandleEnemyDeath;
+    }
+
+    // Fires the spear in the rotation of the gun with its velocity
+    public void Fire(HarpoonGun gun) {
+        this.dropped = false;
+        this.gun = gun;
+
+        Rigidbody2D.velocity = gun.transform.right * gun.harpoonVelocity;
```
Comment (anjx123) [inline Assets/Scripts/Weapon/HarpoonGun/HarpoonSpear.cs:44]: Wait, does right work when the gun is flipped?

File: Assets/Scripts/Weapon/HarpoonGun/HarpoonGun.cs (hunk 4 above)
Comment (arthu-rguo) [reply Assets/Scripts/Weapon/HarpoonGun/HarpoonGun.cs:37]: the LINQer

File: Assets/Scripts/Weapon/HarpoonGun/HarpoonSpear.cs (hunk 5 above)
Comment (arthu-rguo) [reply Assets/Scripts/Weapon/HarpoonGun/HarpoonSpear.cs:44]: I think so. If you think about the angle of the gun as your mouse rotates counter-clockwise, when the player turns from right to left the sprite of the gun actually flips on the Y axis, so "right" always points towards the end of the gun.

**Review by drew-gnaw — APPROVED** [review]

**Pull request merged**